### PR TITLE
fix: do not refresh items on select

### LINKED ISSUE #4861: Grid::select uses given Item instead of Item in Grid to render
### Description

Using `Grid::select`to select an identical item (equals & hashCode is matching) the Grid uses the wrong item to select the items in an already rendered Grid. 

### Expected outcome

4. The grid selects the forth row and the forth row contains "Field 4" **instead of sending the wrong object to the client to render**.

### Minimal reproducible example

```java

    var grid = new Grid<MyBean>();
    grid.addColumn(MyBean::getField).setHeader("Field");
    grid.setItems(Arrays.asList(
      new MyBean("1", "Field 1"),
      new MyBean("2", "Field 2"),
      new MyBean("3", "Field 3")
    ));

    grid.select(new MyBean("3", null));

    add(grid);
    add(new Button("Click me", e -> {
      grid.setItems(Arrays.asList(
        new MyBean("1", "Field 1"),
        new MyBean("2", "Field 2"),
        new MyBean("3", "Field 3"),
        new MyBean("4", "Field 4")
      ));

      grid.select(new MyBean("4", null));
    }));


  private static class MyBean {

    private final String id;
    private final String field;

    private MyBean(String id, String field) {
      this.id = id;
      this.field = field;
    }

    public String getId() {
      return id;
    }

    public String getField() {
      return field;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {return true;}
      if (o == null || getClass() != o.getClass()) {return false;}
      MyBean myBean = (MyBean) o;
      return id.equals(myBean.id);
    }

    @Override
    public int hashCode() {
      return Objects.hash(id);
    }
  }

```

### Steps to reproduce

1. Open a View with the given Code
2. See that `Field 3` is selected
3. Click the Button
4. See that an empty row is created and selected

### Environment

Vaadin version(s): 23.3

### Browsers

Issue is not browser related

### PR DESCRIPTION
**Pull request by ugur-vaadin**
## Description

When an item is selected via `Grid.select`, the provided item is cached in the selection model. However, that item is not necessarily the same item as in the grid. The ID match is enough for that item to be considered for selection. Then, the item is refreshed using the object in this cache. This can lead to incorrect data being sent to the client with `updateFlatData` for the regular `Grid`. 

This PR removes the refresh calls on both single and multi selection models. For the client-side `selected` property state of an item to be consistent, the property is now updated for the server-side selection/deselection calls as well.

Fixes #4861 

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

### CODE REVIEW

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdIT.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        GridElement grid = $(GridElement.class).waitForFirst();
+
+        deselectItemButton.click();
+
+        Assert.assertEquals("1", grid.getCell(0, 1).getText());
+    }
+
+    @Test
+    public void setMultiSelect_addGridWithSelection_selectAnotherItem_itemIsNotUpdated() {
+        useMultiSelectCheckbox.setChecked(true);
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdIT.java:99]: Is it essential to set the multi-select mode before adding the grid to the DOM?

Comment (ugur-vaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdIT.java:99]: The selected item is cached in the selection model. Therefore, the order actually plays a role in the end result. For example, calling `refreshAll` afterwards loses the selection that was made before the selection mode change.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            e.getSource().setVisible(false);
+            add(grid);
+        });
+        addGridButton.setId("add-grid-button");
+
+        add(useMultiSelectCheckbox, addGridButton, selectAnotherItemButton,
+                deselectItemButton);
+    }
+
+    private record Item(String id, String displayValue) {
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java:63]: The `record` modifier is only available in Java 14. Maybe we should use a regular class here to support Java 11.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                                ? Grid.SelectionMode.MULTI
+                                : Grid.SelectionMode.SINGLE));
+        useMultiSelectCheckbox.setId("use-multi-select-checkbox");
+
+        var selectAnotherItemButton = new Button("Select another item",
+                e -> grid.select(new Item("2", "INVALID")));
+        selectAnotherItemButton.setId("select-another-item-button");
+
+        var deselectItemButton = new Button("Deselect item", e -> {
+            System.out.println("Page.deselect");
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java:47]:
```suggestion
```

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import com.vaadin.flow.router.Route;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+@Route("vaadin-grid/select-item-with-identical-id")
+public class SelectItemWithIdenticalIdPage extends Div {
+
+    public SelectItemWithIdenticalIdPage() {
+        var grid = new Grid<Item>();
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java:31]: Let's prefer explicit types to `var`.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        grid.setItems(Arrays.asList(new Item("1", "1"), new Item("2", "2")));
+
+        var useMultiSelectCheckbox = new Checkbox("Use multi-select",
+                event -> grid
+                        .setSelectionMode(Boolean.TRUE.equals(event.getValue())
+                                ? Grid.SelectionMode.MULTI
+                                : Grid.SelectionMode.SINGLE));
+        useMultiSelectCheckbox.setId("use-multi-select-checkbox");
+
+        var selectAnotherItemButton = new Button("Select another item",
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java:42]:
```suggestion
        var selectAnotherItemButton = new Button("Select item 2",
```

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                e -> grid.select(new Item("2", "INVALID")));
+        selectAnotherItemButton.setId("select-another-item-button");
+
+        var deselectItemButton = new Button("Deselect item", e -> {
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java:46]:
```suggestion
        var deselectItemButton = new Button("Deselect item 1", e -> {
```

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            System.out.println("Page.deselect");
+            grid.deselect(new Item("1", "INVALID"));
+        });
+        deselectItemButton.setId("deselect-item-button");
+
+        var addGridButton = new Button("Add grid", e -> {
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdPage.java:52]:
```suggestion
        var addGridWithPreselectionButton = new Button("Add grid", e -> {
```

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdIT.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    private WebElement selectAnotherItemButton;
+
+    private WebElement deselectItemButton;
+
+    private CheckboxElement useMultiSelectCheckbox;
+
+    @Before
+    public void init() {
+        open();
+        addGridButton = findElement(By.id("add-grid-button"));
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/SelectItemWithIdenticalIdIT.java:42]: How about we use the TestBench API for the sake of consistency?

**Review by vursen — COMMENTED** [review]
The solution looks good to me. Only added some minor comments.

**Review by vursen — APPROVED** [review]

**Pull request merged**